### PR TITLE
Int64 encoding enhancements

### PR DESCRIPTION
**Pull request by jwilder**
This PR improves the compression for int64 fields in some special cases.  

The first change is to delta-encode the values before packing them.  This is similar to how the timestamp  encoding works.  This mainly improves counter type fields that increment using relatively small values.  Instead of storing the the potentially large count value at each point, we just store the delta from the prior count.  Small counters values still see some improvement, but not as much as counts with very large values.

The second change is to add run-length encoding when the deltas are all the same.  This can really help values that don't change for long periods of time (e.g. a counter that is usually 0).  

These changes are not backwards compatible with the current format.  Using these new encodings with old values will decode them incorrectly.

### CODE REVIEW

File: tsdb/engine/tsm1/int.go [hunk 1]
```diff
@@ -70,23 +88,56 @@ func (e *int64Encoder) Bytes() ([]byte, error) {
...
+	return b[:i], nil
+}
+
 func (e *int64Encoder) encodePacked() ([]byte, error) {
-	encoded, err := simple8b.EncodeAll(e.values)
+	if len(e.values) == 0 {
+		return nil, nil
+	}
+
+	// Encode all but the first value.  Fist value is written unecoded
```
Comment (otoolep) [inline tsdb/engine/tsm1/int.go]: Why is this change happening?

File: tsdb/engine/tsm1/int.go [hunk 2]
```diff
@@ -151,7 +212,48 @@ func (d *int64Decoder) Error() error {
...
+	// Next 1-10 bytes is the delta value
+	value, n := binary.Uvarint(d.bytes[i:])
+
+	i += n
+
+	// Last 1-10 bytes is how many times the value repeats
+	count, n := binary.Uvarint(d.bytes[i:])
+
+	// Store the first value and delta value so we do not need to allocate
+	// a large values slice.  We can comput the value at position d.i on
```
Comment (otoolep) [inline tsdb/engine/tsm1/int.go]: comput -> compute

File: tsdb/engine/tsm1/int.go (hunk 1 above)
Comment (otoolep) [inline tsdb/engine/tsm1/int.go]: If this is something to with RLE, I don't see the interaction between this code and RLE.

Comment (jwilder) [inline tsdb/engine/tsm1/int.go]: For consistency with the other encoders.  They all check that there are values to encode before trying to encode them.

Comment (otoolep) [inline tsdb/engine/tsm1/int.go]: What I mean is the code used to encode all values, now it just encodes all but the first.

Comment (jwilder) [inline tsdb/engine/tsm1/int.go]: No.  That is due delta-encoding change in 014c51f. It now always delta-encodes the values.  We store the first uncompressed and all the deltas compressed.

Comment (jwilder) [inline tsdb/engine/tsm1/int.go]: First value is written uncompressed here: https://github.com/influxdb/influxdb/pull/4451/files#diff-d0bfc79b383ad5679e03fff72281ee83R127

Comment (otoolep) [inline tsdb/engine/tsm1/int.go]: unecoded -> unencoded

Comment (otoolep) [inline tsdb/engine/tsm1/int.go]: OK, I see it now.

**Comment by otoolep** [comment]
Makes sense, +1.

**Pull request merged**